app/controllers/collections_controller.rb

```diff
@@ -91,11 +91,14 @@ def download_all
   end
 
   def share
-    if @collection.users.exclude?(share_message.recipient) && share_message.save
-      redirect_to collection_path(@collection), notice: t('.success_notice')
-    else
-      redirect_to collection_path(@collection), alert: t('common.errors.something_went_wrong')
-    end
+    flash_message = if @collection.users.include?(share_message.recipient) || Message.exists?(share_message_args)
+                      {alert: t('.duplicate_share')}
+                    elsif share_message.save
+                      {notice: t('.success_notice')}
+                    else
+                      {alert: t('common.errors.something_went_wrong')}
+                    end
+    redirect_to collection_path(@collection), flash_message
```

Comment:
I would prefer to have the message validation in the model, where business logic should be placed. With a proper backend validation, we could simply write:

```suggestion
    flash_message = if @collection.users.exclude?(share_message.recipient) && share_message.save
                      {notice: t('.success_notice')}
                    else
                      {alert: share_message.errors.full_messages.join(', ')}
                    end
    redirect_to collection_path(@collection), flash_message
```

I am currently not sure whether we have this kind of validation already, since the following line looks already suitable:

https://github.com/openHPI/codeharbor/blob/5dac5a122eab1e6d66199d07ff349658f1e7c03e/app/models/message.rb#L5

We can modify this line slightly:

```rb
  validates :recipient_id, uniqueness: {scope: %i[param_id param_type], message: :duplicate_share}, if: -> { param_type == 'collection' }
```

then, in the translations, we just use:

```yml
en:
  activerecord:
    errors:
      models:
        message:
          duplicate_share: has already been added to this collection. # TODO @Mathis-Z: Or some other nice text.
```

I feel this would be more Rails-like to handle this error.

Reply:
Since some tests are failing due to the missing changes for the `Messages` model (and the error translations), I am reopening this thread. I would assume that the changes proposed by me earlier should resolve the test failures.

Reply:
Sorry, I was not done yet.

Reply:
Now I am done :)